group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -601,6 +601,33 @@ ConsumerGroup getOrMaybeCreateConsumerGroup(
             throw new GroupIdNotFoundException(String.format("Consumer group %s not found.", groupId));
         }
 
+        if (group == null) {
+            ConsumerGroup consumerGroup = new ConsumerGroup(snapshotRegistry, groupId, metrics);
+            return consumerGroup;
+        } else {
+            if (group.type() == CONSUMER) {
+                return (ConsumerGroup) group;
+            } else {
+                // We don't support upgrading/downgrading between protocols at the moment so
+                // we throw an exception if a group exists with the wrong type.
+                throw new GroupIdNotFoundException(String.format("Group %s is not a consumer group.", groupId));
+            }
+        }
+    }
+
+    /**
+     * Gets or maybe creates a consumer group. Updates the groups map if a new group is created.
+     *
+     * @param groupId           The group id.
+     *
+     * @return A ConsumerGroup.
+     * @throws GroupIdNotFoundException if the group does not exist and createIfNotExists is false or
+     *                                  if the group is not a consumer group.
+     */
+    private ConsumerGroup maybeCreateAndUpdateConsumerGroup(
```

Comment:
The naming is a bit confusing here. I wonder if we could use something like `getOrMaybeCreatePersistedConsumerGroup` to denote that we are persisting the group in this one. We may also be able to keep the `createIfNotExists`.